pkg/whiteflag/test/whiteflag_test.go

```diff
@@ -0,0 +1,599 @@
+	require.NotNil(t, coo)
+
+	err = coo.InitMerkleTree("coordinator.tree", cooAddress)
+	require.NoError(t, err)
+
+	coo.InitState(true, 0)
+
+	// Save snapshot info
+	tangle.SetSnapshotMilestone(hornet.HashFromAddressTrytes(cooAddress), hornet.NullHashBytes, 0, 0, 0, time.Now().Unix(), false)
+
+	// Configure Milestones
+	tangle.ConfigureMilestones(hornet.HashFromAddressTrytes(cooAddress), int(secLevel), merkleTreeDepth, merkleHashFunc)
+
+	return coo
+}
+
+func verifyLSMI(t *testing.T, index milestone.Index) {
+	lsmi := tangle.GetSolidMilestoneIndex()
+	require.Equal(t, index, lsmi)
+}
+
+func verifyLMI(t *testing.T, index milestone.Index) {
+	lmi := tangle.GetLatestMilestoneIndex()
+	require.Equal(t, index, lmi)
+}
+
+func issueAndConfirmMilestoneOnTip(t *testing.T, tip hornet.Hash, printTangle bool) (*tangle.CachedBundle, *whiteflag.ConfirmedMilestoneStats) {
+
+	nextTip = tip
+
+	currentIndex := tangle.GetSolidMilestoneIndex()
+	verifyLMI(t, currentIndex)
+
+	fmt.Printf("Issue milestone %v\n", currentIndex+1)
+	noncriticalErr, criticalErr := coo.IssueNextCheckpointOrMilestone()
+	require.NoError(t, noncriticalErr)
+	require.NoError(t, criticalErr)
+
+	verifyLMI(t, currentIndex+1)
+
+	milestoneIndex := currentIndex + 1
+	ms := tangle.GetMilestoneOrNil(milestoneIndex)
+	require.NotNil(t, ms)
+
+	conf, err := whiteflag.ConfirmMilestone(ms.Retain(), func(tx *tangle.CachedTransaction, index milestone.Index, confTime int64) {}, func(confirmation *whiteflag.Confirmation) {
+		tangle.SetSolidMilestoneIndex(confirmation.MilestoneIndex)
+	})
+	require.NoError(t, err)
+
+	require.Equal(t, currentIndex+1, conf.Index)
+	verifyLSMI(t, conf.Index)
+
+	cachedBundles = append(cachedBundles, ms)
+
+	assertTotalSupplyStillValid(t)
+
+	if printTangle {
+		fmt.Print(generateDotFileFromTangle(t))
+	}
+
+	return ms, conf
+}
+
+func generateAddress(t *testing.T, seed trinary.Trytes, index uint64) hornet.Hash {
+	seedAddress, err := address.GenerateAddress(seed, index, consts.SecurityLevelMedium, false)
+	require.NoError(t, err)
+	return hornet.HashFromAddressTrytes(seedAddress)
+}
+
+func assertAddressBalance(t *testing.T, seed trinary.Trytes, index uint64, balance uint64) {
+
+	address := generateAddress(t, seed, index)
+	addrBalance, _, err := tangle.GetBalanceForAddress(address)
+	require.NoError(t, err)
+	require.Equal(t, balance, addrBalance)
+}
+
+func assertTotalSupplyStillValid(t *testing.T) {
+	_, _, err := tangle.GetLedgerStateForLSMI(nil)
+	require.NoError(t, err)
+}
+
+func setupCoordinatorAndIssueInitialMilestones(t *testing.T, initialBalances map[string]uint64, numberOfMilestones int) tangle.CachedBundles {
+
+	balances := initialBalances
+	var sum uint64
+	for _, value := range balances {
+		sum += value
+	}
+	// Move remaining supply to 999..999
+	balances[string(hornet.NullHashBytes)] = consts.TotalSupply - sum
+
+	store := mapdb.NewMapDB()
+	setupTestEnvironment(t, store)
+
+	solidEntryPointsOnce.Do(func() {
```

Comment:
Why is it called `solidEntryPointsOnce` ?

Reply:
forgot to rename it at the end, but actually it protects that we only itinialize the solid entry points once. We have some global state there that I cannot reset betweeen each test. I'd need to adapt the code, but we made it so to actually catch wrong usage. I'll rename the once